File: src/java/com/kitware/pulse/cdm/properties/SESegment.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.properties;

import com.kitware.pulse.cdm.bind.Properties.AnySegmentData;

public class SESegment
{
  protected SEScalarVolume beginVolume;
  protected SEScalarVolume endVolume;

  protected SESegment()
  {
    beginVolume = null;
    endVolume = null;
  }

  public SESegment(SESegment other)
  {
    copy(other);
  }

  public void copy(SESegment other)
  {
    if (other.hasBeginVolume())
      getBeginVolume().set(other.getBeginVolume());
    if (other.hasEndVolume())
      getEndVolume().set(other.getEndVolume());
  }

  public void clear()
  {
    if (beginVolume != null)
      beginVolume.invalidate();
    if (endVolume != null)
      endVolume.invalidate();
  }

  public boolean isValid()
  {
    return hasBeginVolume() && hasEndVolume();
  }

  public boolean hasBeginVolume()
  {
    return beginVolume == null ? false : beginVolume.isValid();
  }

  public SEScalarVolume getBeginVolume()
  {
    if (beginVolume == null)
      beginVolume = new SEScalarVolume();
    return beginVolume;
  }

  public boolean hasEndVolume()
  {
    return endVolume == null ? false : endVolume.isValid();
  }

  public SEScalarVolume getEndVolume()
  {
    if (endVolume == null)
      endVolume = new SEScalarVolume();
    return endVolume;
  }

  public String toString()
  {
    String str = "";
    str += "\n\tBeginVolume: " + (hasBeginVolume() ? getBeginVolume().toString() : "Not Provided");
    str += "\n\tEndVolume: " + (hasEndVolume() ? getEndVolume().toString() : "Not Provided");

    return str;
  }
}

File: src/csharp/pulse/cdm/properties/SESegment.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SESegment
  {
    protected SEScalarVolume beginVolume;
    protected SEScalarVolume endVolume;

    protected SESegment()
    {
      beginVolume = null;
      endVolume = null;
    }

    public SESegment(SESegment other)
    {
      Copy(other);
    }

    public void Copy(SESegment other)
    {
      if (other.HasBeginVolume())
        GetBeginVolume().Set(other.GetBeginVolume());
      if (other.HasEndVolume())
        GetEndVolume().Set(other.GetEndVolume());
    }

    public virtual void Clear()
    {
      if (beginVolume != null)
        beginVolume.Invalidate();
      if (endVolume != null)
        endVolume.Invalidate();
    }

    public virtual bool IsValid()
    {
      return HasBeginVolume() && HasEndVolume();
    }

    public bool HasBeginVolume()
    {
      return beginVolume == null ? false : beginVolume.IsValid();
    }
    public SEScalarVolume GetBeginVolume()
    {
      if (beginVolume == null)
        beginVolume = new SEScalarVolume();
      return beginVolume;
    }

    public bool HasEndVolume()
    {
      return endVolume == null ? false : endVolume.IsValid();
    }
    public SEScalarVolume GetEndVolume()
    {
      if (endVolume == null)
        endVolume = new SEScalarVolume();
      return endVolume;
    }

    public override string ToString()
    {
      string str="";
      str += "\n\tBeginVolume: " + (HasBeginVolume() ? GetBeginVolume().ToString() : "Not Provided");
      str += "\n\tEndVolume: " + (HasEndVolume() ? GetEndVolume().ToString() : "Not Provided");

      return str;
    }
  }
}
